src/codegen.c

```diff
@@ -357,34 +357,46 @@ static void codegen_expr(CodegenContext *ctx, Node *expr) {
         return;
       }
 
-      /// Anything else is an error (I think?).
-      /// Otherwise, actually emit the LHS and load from that.
-      /*codegen_expr(ctx, lhs);
-      expr->ir = ir_store(ctx, lhs->ir, rhs->ir);*/
-      return;
+      ICE("Invalid type of lhs of assignment (should have been caught by sema): %T", lhs->type);
     }
 
-    // TODO: Just use lhs operand of subscript operator when right hand
-    // side is a compile-time-known zero value.
-
-    /// Emit the operands.
-    codegen_expr(ctx, lhs);
-    codegen_expr(ctx, rhs);
     if (expr->binary.op == TK_LBRACK) {
+      codegen_expr(ctx, rhs);
+
+      IRInstruction *subs_lhs = NULL;
+      if (lhs->type->kind != TYPE_ARRAY && lhs->type->kind == TYPE_POINTER) {
+        ERR("Subscript operator may only operate on arrays and pointers, which type %T is not", lhs->type);
+      }
+      if (lhs->kind == NODE_VARIABLE_REFERENCE) {
+        // TODO: Handle local variable references, somehow. How can we tell if it's local/static?
```

Comment:
![image](https://user-images.githubusercontent.com/69637718/217108632-9dbe91b8-f71b-4e66-9ea8-24fdf17ae7bf.png)

There **aren't** any `STATIC_REF` nodes in the AST... ???

Reply:
Er... I see my bad. In that case there is probably a `decl` member in every `NODE_VARIABLE_REFERENCE` from which you can get the `NODE_DECLARATION` for that variable, and it should contain a reference to an `IRInstruction` that is either an `IRStaticVariable` or an `IRAlloca` iirc. It’s been a while since I last looked into the inner workings of this AST/IR, but seeing as I wrote part of it, that is how I’d do it now.

Reply:
So the symbol will have a reference to the IRInstruction we need?
https://github.com/LensPlaysGames/FUNCompiler/blob/33c195c563a700315e6e5932ddfea2b98f14bcc6/src/ast.h#L225

Reply:
Right there’s an extra step. The varref contains a symbol, the symbol contains a `Node*` that has to be a variable declaration, the declaration contains an `IRInstruction*` that is either an alloca or static var

Reply:
Or rather, a varref is just a symbol

Reply:
What instruction should locals generate? Creating another `alloca` doesn't feel right

Reply:
Wdym? A local variable *is* an alloca

Reply:
Not a reference to one, though.

i.e.

```
foo : integer ;; an alloca
foo ;; a reference to an alloca (what do we codegen here in local context?)
```

Reply:
With static refs, we just create more static refs

Reply:
Currently, a variable reference is a `load`. Like I mentioned earlier, we might want to have both an `.ir` and `.ir_addr` member or sth like that at some point

Reply:
```c
        IRInstruction *var = lhs->var->val.node->ir;
        // ASSERT(var);
        if (var->kind == IR_STATIC_REF)
          subs_lhs = ir_static_reference(ctx, as_span(lhs->var->name));
        else if (var->kind == IR_ALLOCA)
          //subs_lhs = ir_(ctx, as_span(lhs->var->name));
          TODO("IR generation for subscript of local variable reference");
 ```
Basically, this is the code now. We need to somehow use the alloca; can we just use it directly?

Reply:
Well, if you have an alloca then you can use it directly yeah. If you look at the codegen for `*a := 4` or whatever (that is, codegen for assignment exprs), we’re pretty much doing exactly that.

Reply:
And by ‘using the alloca’ directly I mean, you can just use the IRInstruction that is the alloca

Reply:
It doesn't work, sadly. I get random results. Seemingly, the IR looks "fine".

371397897
-364112631
-206760695
etc...

![image](https://user-images.githubusercontent.com/69637718/217161957-159f926a-3784-4a69-afba-1948f87d2e59.png)

Reply:
The IR looks fine yeah... is there anything sus in the ASM maybe? :eyes:

Reply:
Basically; while this works in the IR, the two address nature of x86_64 makes it really hard to keep around a value like this

Reply:
```assembly
_XF3fooF7integerE:
    push %rbp
    mov %rsp, %rbp
    sub $88, %rsp
.L1:
    lea -32(%rbp), %rax
    lea -40(%rbp), %rcx
    mov $0, %edx
    mov $8, %r8
    imul %rdx, %r8
    mov %r8, %rdx
    add %rdx, %rax
    mov %rax, %rdx
    mov %rdx, -40(%rbp)
    lea -48(%rbp), %rcx
    mov $0, %edx
    mov $8, %r8
    imul %rdx, %r8
    mov %r8, %rdx
    add %rdx, %rax
    mov %rax, %rdx
    mov 0(%rdx), %rdx
    mov %rdx, -48(%rbp)
    mov $69, %ecx
    mov $0, %edx
    mov $8, %r8
    imul %rdx, %r8
    mov %r8, %rdx
    add %rdx, %rax
    mov %rax, %rdx
    mov %rcx, (%rdx)
    mov $420, %ecx
    mov $1, %edx
    mov $8, %r8
    imul %rdx, %r8
    mov %r8, %rdx
    add %rdx, %rax
    mov %rax, %rdx
    mov %rcx, (%rdx)
    mov $69, %ecx
    mov $2, %edx
    mov $8, %r8
    imul %rdx, %r8
    mov %r8, %rdx
    add %rdx, %rax
    mov %rax, %rdx
    mov %rcx, (%rdx)
    mov $420, %ecx
    mov $3, %edx
    mov $8, %r8
    imul %rdx, %r8
    mov %r8, %rdx
    add %rdx, %rax
    mov %rax, %rdx
    mov %rcx, (%rdx)
    mov $2, %ecx
    mov $8, %rdx
    imul %rcx, %rdx
    mov %rdx, %rcx
    add %rcx, %rax
    mov 0(%rax), %rax
    mov %rbp, %rsp
    pop %rbp
    ret
```

Reply:
As you can see, each time the subscript happens and the result is in `rax`, it gets overwritten and never reset

Reply:
> Basically; while this works in the IR, the two address nature of x86_64 makes it really hard to keep around a value like this

Well that’s a failure in the RA/backend then. Instead of
```assembly
    add %rdx, %rax
    mov %rax, %rdx
```
we should be emitting
```assembly
    add %rax, %rdx
```

Reply:
True; looks like I need to delve into RA at some point to solve this; I may make an issue tomorrow and not include that in this PR, for the sake of both time and size.

Reply:
The relevant code snippets are these:
https://github.com/LensPlaysGames/FUNCompiler/blob/f3fab2d983e4caeba7e9a47c39292bb54508e78c/src/codegen/x86_64/arch_x86_64.c#L930-L932

https://github.com/LensPlaysGames/FUNCompiler/blob/f3fab2d983e4caeba7e9a47c39292bb54508e78c/src/codegen/x86_64/arch_x86_64.c#L1164-L1173

https://github.com/LensPlaysGames/FUNCompiler/blob/f3fab2d983e4caeba7e9a47c39292bb54508e78c/src/codegen/x86_64/arch_x86_64.c#L402

Reply:
We’re emitting the IR instruction`add lhs, rhs` as the x86_64 instruction `add %rhs, %lhs`. These are the wrong way round it seems

Reply:
So in sum, the problem is that we’re telling the RA that the RHS is clobbered, but we’re clobbering the LHS

Reply:
Interesting; so solving the IR through subscript generation parameter order flipping may be the way to go after all

Reply:
:O 

it worked

Reply:
The problem is in the backend. This
```c
femit_reg_to_reg(context, I_ADD, inst->rhs->result, inst->lhs->result); 
```
should instead be
```c
femit_reg_to_reg(context, I_ADD, inst->lhs->result, inst->rhs->result); 
```

Reply:
a8241fa4a48c6a63bf09a570e82cbbe95f61994f

Reply:
That or we say that it CLOBBERS_LEFT. I think the reason why this part of the backend/RA is as confused as it is is because I think in Intel syntax, and you in AT&T syntax, so we end up having a different understanding of what’s the dest and what the source...

Reply:
Hahaha, that's definitely true. That's why we need explicit commutative assignment KEKW

`a := b` == `b := a`

Reply:
> :O
> 
> it worked

It may have, but you’re fixing the symptoms, not the problem. IR_ADD is still fundamentally broken. The order in which you add them ought not to matter because ADD is a commutative operation. This is a problem w/ the backend

Reply:
How would it be fixed? Converting three address to two address is really hard lmao

Reply:
> The problem is in the backend. This
> 
> ```c
> femit_reg_to_reg(context, I_ADD, inst->rhs->result, inst->lhs->result); 
> ```
> 
> should instead be
> 
> ```c
> femit_reg_to_reg(context, I_ADD, inst->lhs->result, inst->rhs->result); 
> ```

Like so

Reply:
Oh. I thought you meant that *even with those changes*, IR_ADD is still fundamentally broken.

6c027c7e7928ddc406bb8440ac9a5e1270cdff58

Reply:
And make sure that the CLOBBERS_LEFT/CLOBBERS_RIGHT values for all binary instructions match the order in which their LHS and RHS are emitted here. Not all instructions (e.g. sub) are commutative, so we might have to change CLOBBERS_LEFT to CLOBBERS_RIGHT or vice versa for some instructions depending if there are any other instructions that are similarly messed up

Reply:
> Oh. I thought you meant that _even with those changes_, IR_ADD is still fundamentally broken.
> 
> [6c027c7](https://github.com/LensPlaysGames/FUNCompiler/commit/6c027c7e7928ddc406bb8440ac9a5e1270cdff58)

No, w/ those changes, it *should* be fixed, unless I’m missing sth ;Þ

Reply:
True; it looks like `SAR` and friends *might* be messed up? It's hard to tell exactly (it's late right now, haha).

```c
    femit_reg_to_reg(context, I_MOV, inst->rhs->result, REG_RCX);
    femit_reg(context, I_SAR, inst->lhs->result);
    femit_reg_to_reg(context, I_MOV, inst->lhs->result, inst->result);
```

This would clobber ... left hand side, right?

Reply:
> This would clobber ... left hand side, right?

Yup. As well as RCX ofc.

Reply:
> Yup.

F

So this is definitely a problem right now haha
```c
  case IR_SHL:
  case IR_SHR:
  case IR_SAR:
  case IR_AND:
  case IR_OR:
    return CLOBBERS_RIGHT;
```

Reply:
*possibly*

Reply:
Fixed in c7f87aac0c465ef339c5ec36519ef210ecd53767